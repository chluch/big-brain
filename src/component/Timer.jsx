import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

// Based off https://css-tricks.com/how-to-create-an-animated-countdown-timer-with-html-css-and-javascript/
// but using React, no external CSS

const formatNumXPosition = (seconds) => {
  if (seconds < 10) {
    return 42;
  }
  if (seconds > 99) {
    return 29;
  }
  return 35;
};

// Why 283?
// circumference of circle = 2πr
// svg r=45, therefore inital circleDasharray = 2π(45) ~= 283
// If we change radius of svg we need to recalculate circleDasharray

const Timer = ({ question, setTimesUp }) => {
  const [time, setTime] = useState(question.timeLimit);
  const [circleDasharray, setCircleDasharray] = useState(283);
  const [colour, setColour] = useState('#21ada8');
  useEffect(() => {
    setTime(question.timeLimit);
  }, [question]);

  useEffect(() => {
    let countdownTimer;
    setCircleDasharray(`${(time / question.timeLimit) * 283} 283`);
    setColour('#21ada8');
    if (time <= question.timeLimit / 2) {
      setColour('orange');
    }
    if (time <= question.timeLimit / 3) {
      setColour('red');
    }
    if (time > 0) {
      countdownTimer = setTimeout(() => setTime(time - 1), 1000);
    }
    if (time === 0) {
      setTimesUp(true);
    }
    return () => {
      clearTimeout(countdownTimer);
    };
  }, [setTimesUp, time, question.timeLimit]);

  return (
    <div
      style={{
        fontSize: '24px',
      }}
    >
      {/* Size can be adjusted by height and width here */}
      <svg
        className="timer"
        viewBox="0 0 100 100"
        height="100"
        width="100"
        style={{ transform: 'scaleX(-1)' }}
      >
        <g>
          <circle cx="50" cy="50" r="45" fill="none" stroke="#333" strokeWidth="9px" />
          <text
            x={formatNumXPosition(time)}
            y="60"
            fill="black"
            style={{
              transform: 'scaleX(-1)',
              transformOrigin: 'center',
            }}
          >
            {time}
          </text>
          <path
            id="base-timer-path-remaining"
            stroke={colour}
            fill="none"
            strokeWidth="9px"
            strokeDasharray={circleDasharray}
            style={{
              transform: 'rotate(90deg)',
              transformOrigin: 'center',
            }}
            d="M 50, 50
              m -45, 0
              a 45,45 0 1,0 90,0
              a 45,45 0 1,0 -90,0"
          />
        </g>
      </svg>
    </div>
  );
};

Timer.propTypes = {
  question: PropTypes.shape({
    timeLimit: PropTypes.number.isRequired,
  }).isRequired,
  setTimesUp: PropTypes.func.isRequired,
};

export default Timer;
