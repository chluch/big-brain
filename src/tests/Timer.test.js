import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer, { act } from 'react-test-renderer';
import Timer from '../component/Timer';

const data1 = {
  timeLimit: 10,
};
const data2 = {
  timeLimit: 30,
};

it('renders a time limit', () => {
  const wrapper = shallow(<Timer question={data1} setTimesUp={() => { }} />);
  const time = wrapper.find('text');
  expect(parseInt(time.text(), 10)).toEqual(data1.timeLimit);
});

it('renders an unfilled circle with stroke and stroke width defined', () => {
  const wrapper = shallow(<Timer question={data1} setTimesUp={() => { }} />);
  const circle = wrapper.find('circle');
  expect(circle.length).toBe(1);
  expect(circle.props().fill).toEqual('none');
  expect(circle.props().stroke).not.toEqual('none');
  expect(circle.props().strokeWidth).toBeDefined();
});

it('renders a path that is coloured #21ada8', () => {
  const wrapper = shallow(<Timer question={data1} setTimesUp={() => { }} />);
  const path = wrapper.find('path');
  expect(path.length).toBe(1);
  expect(path.props().stroke).toEqual('#21ada8');
});

it('decrements the time by 1 after 1 second passed', () => {
  jest.useFakeTimers();
  const wrapper = mount(<Timer question={data1} setTimesUp={() => { }} />);
  let time = wrapper.find('text');
  expect(parseInt(time.text(), 10)).toEqual(data1.timeLimit);
  act(() => {
    jest.advanceTimersByTime(1000);
    wrapper.update();
    time = wrapper.find('text');
    expect(parseInt(time.text(), 10)).toEqual(data1.timeLimit - 1);
  });
});

it('Stops decrementing when countdown reaches 0', () => {
  jest.useFakeTimers();
  const wrapper = mount(<Timer question={data1} setTimesUp={() => { }} />);
  let time = wrapper.find('text');
  expect(parseInt(time.text(), 10)).toEqual(data1.timeLimit);
  act(() => {
    jest.advanceTimersByTime(20000);
    wrapper.update();
    time = wrapper.find('text');
    expect(parseInt(time.text(), 10)).toEqual(0);
  });
});

it('changes colour to orange when countdown reaches the 50% mark', () => {
  jest.useFakeTimers();
  const wrapper = mount(<Timer question={data1} setTimesUp={() => { }} />);
  let svg = wrapper.find('path');
  expect(svg.props().stroke).toEqual('#21ada8');
  act(() => {
    jest.advanceTimersByTime(5000);
    wrapper.update();
    svg = wrapper.find('path');
    expect(svg.props().stroke).toEqual('orange');
  });
});

it('changes colour to red when countdown passes the 33% mark (1/3 time left)', () => {
  jest.useFakeTimers();
  const wrapper = mount(<Timer question={data2} setTimesUp={() => { }} />);
  let svg = wrapper.find('path');
  expect(svg.props().stroke).toEqual('#21ada8');
  act(() => {
    jest.advanceTimersByTime(20000);
    wrapper.update();
    svg = wrapper.find('path');
    expect(svg.props().stroke).toEqual('red');
  });
});

it('calls setTimesUp when countdown finishes', () => {
  jest.useFakeTimers();
  const mockFn = jest.fn();
  const wrapper = mount(<Timer question={data1} setTimesUp={mockFn} />);
  const time = wrapper.find('text');
  expect(mockFn).not.toBeCalled();
  act(() => {
    jest.runAllTimers();
    expect(mockFn).toHaveBeenCalledTimes(1);
    expect(parseInt(time.text(), 10)).toEqual(0);
  });
});

// Snapshot
it('renders correctly when counting down each second', () => {
  jest.useFakeTimers();
  const timer = renderer.create(<Timer question={data1} setTimesUp={() => { }} />);
  act(() => {
    jest.advanceTimersByTime(1000);
    expect(timer).toMatchSnapshot();
    jest.advanceTimersByTime(1000);
    expect(timer).toMatchSnapshot();
  });
});
