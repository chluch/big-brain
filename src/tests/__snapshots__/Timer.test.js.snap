// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when counting down each second 1`] = `
<div
  style={
    Object {
      "fontSize": "24px",
    }
  }
>
  <svg
    className="timer"
    height="100"
    style={
      Object {
        "transform": "scaleX(-1)",
      }
    }
    viewBox="0 0 100 100"
    width="100"
  >
    <g>
      <circle
        cx="50"
        cy="50"
        fill="none"
        r="45"
        stroke="#333"
        strokeWidth="9px"
      />
      <text
        fill="black"
        style={
          Object {
            "transform": "scaleX(-1)",
            "transformOrigin": "center",
          }
        }
        x={42}
        y="60"
      >
        9
      </text>
      <path
        d="M 50, 50 m -45, 0 a 45,45 0 1,0 90,0 a 45,45 0 1,0 -90,0"
        fill="none"
        id="base-timer-path-remaining"
        stroke="#21ada8"
        strokeDasharray="254.70000000000002 283"
        strokeWidth="9px"
        style={
          Object {
            "transform": "rotate(90deg)",
            "transformOrigin": "center",
          }
        }
      />
    </g>
  </svg>
</div>
`;

exports[`renders correctly when counting down each second 2`] = `
<div
  style={
    Object {
      "fontSize": "24px",
    }
  }
>
  <svg
    className="timer"
    height="100"
    style={
      Object {
        "transform": "scaleX(-1)",
      }
    }
    viewBox="0 0 100 100"
    width="100"
  >
    <g>
      <circle
        cx="50"
        cy="50"
        fill="none"
        r="45"
        stroke="#333"
        strokeWidth="9px"
      />
      <text
        fill="black"
        style={
          Object {
            "transform": "scaleX(-1)",
            "transformOrigin": "center",
          }
        }
        x={42}
        y="60"
      >
        8
      </text>
      <path
        d="M 50, 50 m -45, 0 a 45,45 0 1,0 90,0 a 45,45 0 1,0 -90,0"
        fill="none"
        id="base-timer-path-remaining"
        stroke="#21ada8"
        strokeDasharray="226.4 283"
        strokeWidth="9px"
        style={
          Object {
            "transform": "rotate(90deg)",
            "transformOrigin": "center",
          }
        }
      />
    </g>
  </svg>
</div>
`;
